File: Textbook_Exercises/Python/Chapter_9/Exercise_9_2/PCA_demo.py
# This file is associated with the book
# "Machine Learning Refined", Cambridge University Press, 2016.
# by Jeremy Watt, Reza Borhani, and Aggelos Katsaggelos.

import numpy as np
import numpy.matlib
import matplotlib.pyplot as plt
from __future__ import division

def PCA_demo():

    global n

    # load in data
    X = np.matrix(np.genfromtxt('PCA_demo_data.csv', delimiter=','))
    n = np.shape(X)[0]
    means = np.matlib.repmat(np.mean(X,0), n, 1)
    X = X - means  # center the data
    X = X.T
    K = 1

    # run PCA
    C, W = your_PCA(X, K)

    # plot results
    plot_results(X, C)
    plt.show()

    return


def your_PCA(X, K):

# ---> YOUR CODE GOES HERE.

    return C, W

def plot_results(X, C):

    # Print points and pcs
    fig = plt.figure(facecolor = 'white')
    ax1 = fig.add_subplot(121)
    for j in np.arange(0,n):
        plt.scatter(X[0][:],X[1][:])

    s = np.arange(C[0,0],-C[0,0],.001)
    m = C[1,0]/C[0,0]
    ax1.plot(s, m*s, color = 'k', linewidth = 2)

    ax1.set_xlabel('$b_1$', fontsize = 14)
    ax1.set_ylabel('$b_2$', fontsize = 14)
    ax1.set_xlim(-.5, .5)
    ax1.set_ylim(-.5, .5)
    ax1.set_aspect('equal')

    # Plot projected data
    ax2 = fig.add_subplot(122)
    X_proj = np.dot(C, np.linalg.solve(np.dot(C.T,C),np.dot(C.T,X)))
    for j in np.arange(0,n):
        plt.scatter(X_proj[0][:],X_proj[1][:])

    ax2.set_xlabel('$b_1$', fontsize = 14)
    ax2.set_ylabel('$b_2$', fontsize = 14)
    ax2.set_xlim(-.5, .5)
    ax2.set_ylim(-.5, .5)
    ax2.set_aspect('equal')

    return

PCA_demo()

File: mlrefined_exercises/ed_1/MATLAB/Chapter_9/Exercise_9_2/PCA_demo.m
function PCA_demo()

% This file is associated with the book
% "Machine Learning Refined", Cambridge University Press, 2016.
% by Jeremy Watt, Reza Borhani, and Aggelos Katsaggelos.

% load in data
X = csvread('PCA_demo_data.csv');
n = size(X,1);
means = repmat(mean(X),n,1);
X = X - means;  % center the data
X = X';

K = 1;  

% run PCA 
[C, W] = your_PCA(X, K);

% plot_results
plot_results(X, C)


function [C, W] = your_PCA(X, K)

% ---->  YOUR CODE GOES HERE    
    
end


function plot_results(X, C)
        
    % Print points and pcs
    subplot(1,2,1)
    for j = 1:n
        hold on
        scatter(X(1,:),X(2,:),'fill','b')
    end
    hold on
    s = C(1,1):0.001:-C(1,1);
    m = C(2,1)/C(1,1);
    plot(s,m*s,'k','LineWidth',2.5)
    xlabel('b_1','Fontsize',14,'FontName','cmr10')
    ylabel('b_2','Fontsize',14,'FontName','cmr10')

    axis([-0.5 0.5 -0.5 0.5])    % Set viewing axes
    axis square


    % Plot projected data
    subplot(1,2,2)

    X_proj = C*((C'*C)\(C'*X));
    for j = 1:n
        hold on
        scatter(X_proj(1,:),X_proj(2,:),'fill','b')
    end
    axis([-0.5 0.5 -0.5 0.5])    % Set viewing axes
    xlabel('b_1','Fontsize',14,'FontName','cmr10')
    ylabel('b_2','Fontsize',14,'FontName','cmr10')

    axis square
    set(gcf,'color','w');

end

end


    
    








